File: backend/api/candidate_activities.py
"""
Candidate activity timeline endpoints.

This module provides endpoints for retrieving candidate activity history,
including stage changes, notes additions/changes, tag modifications,
and other significant candidate events throughout the hiring process.
"""
import logging
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models.candidate_activity import CandidateActivity, CandidateActivityType
from models.resume import Resume

logger = logging.getLogger(__name__)

router = APIRouter()


class ActivityItem(BaseModel):
    """Single activity item in the timeline."""

    id: str = Field(..., description="Activity ID")
    activity_type: str = Field(..., description="Type of activity (e.g., 'stage_changed', 'note_added')")
    candidate_id: str = Field(..., description="Candidate (resume) ID")
    vacancy_id: Optional[str] = Field(None, description="Related vacancy ID if applicable")
    from_stage: Optional[str] = Field(None, description="Previous stage for stage changes")
    to_stage: Optional[str] = Field(None, description="New stage for stage changes")
    note_id: Optional[str] = Field(None, description="Related note ID if applicable")
    tag_id: Optional[str] = Field(None, description="Related tag ID if applicable")
    recruiter_id: Optional[str] = Field(None, description="Recruiter who performed the action")
    activity_data: Optional[dict] = Field(None, description="Additional activity-specific data")
    reason: Optional[str] = Field(None, description="Reason or explanation for the activity")
    created_at: str = Field(..., description="When the activity occurred")


class ActivityTimelineResponse(BaseModel):
    """Response model for candidate activity timeline."""

    resume_id: str = Field(..., description="Resume ID")
    activities: List[ActivityItem] = Field(..., description="List of activities in chronological order")
    total_count: int = Field(..., description="Total number of activities")


@router.get(
    "/",
    response_model=ActivityTimelineResponse,
    tags=["Candidate Activities"],
)
async def get_candidate_activities(
    resume_id: Optional[str] = Query(None, description="Filter by resume (candidate) ID"),
    activity_type: Optional[str] = Query(None, description="Filter by activity type"),
    vacancy_id: Optional[str] = Query(None, description="Filter by vacancy ID"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of activities to return"),
    offset: int = Query(0, ge=0, description="Number of activities to skip for pagination"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Get candidate activity timeline.

    This endpoint retrieves the activity history for a candidate, including
    stage changes, notes additions/changes, tag modifications, and other
    significant events throughout the hiring process.

    Activities are returned in reverse chronological order (newest first).

    Args:
        resume_id: Optional filter to get activities for a specific resume
        activity_type: Optional filter for specific activity type (e.g., 'stage_changed', 'note_added')
        vacancy_id: Optional filter for specific vacancy
        limit: Maximum number of activities to return (default: 100)
        offset: Number of activities to skip for pagination (default: 0)
        db: Database session

    Returns:
        JSON response with list of activities in chronological order

    Raises:
        HTTPException(404): If resume_id is provided and resume not found
        HTTPException(400): If activity_type is invalid
        HTTPException(500): If data retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/candidate-activities/?resume_id=abc-123")
        >>> response.json()
        {
            "resume_id": "abc-123",
            "activities": [
                {
                    "id": "act-1",
                    "activity_type": "stage_changed",
                    "candidate_id": "abc-123",
                    "vacancy_id": "vac-1",
                    "from_stage": "screening",
                    "to_stage": "interview",
                    "note_id": null,
                    "tag_id": null,
                    "recruiter_id": "rec-1",
                    "activity_data": null,
                    "reason": "Candidate passed initial screening",
                    "created_at": "2026-01-31T10:30:00Z"
                }
            ],
            "total_count": 1
        }
    """
    try:
        logger.info(
            f"Fetching candidate activities - resume_id: {resume_id}, "
            f"activity_type: {activity_type}, vacancy_id: {vacancy_id}"
        )

        # Build base query
        query = select(CandidateActivity)

        # Apply filters
        if resume_id:
            # Verify resume exists
            resume_result = await db.execute(
                select(Resume).where(Resume.id == UUID(resume_id))
            )
            resume = resume_result.scalar_one_or_none()

            if not resume:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Resume not found: {resume_id}",
                )

            query = query.where(CandidateActivity.candidate_id == UUID(resume_id))

        if vacancy_id:
            query = query.where(CandidateActivity.vacancy_id == UUID(vacancy_id))

        if activity_type:
            # Validate activity type
            valid_types = [t.value for t in CandidateActivityType]
            if activity_type not in valid_types:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Invalid activity_type: {activity_type}. "
                           f"Valid types are: {', '.join(valid_types)}",
                )
            query = query.where(CandidateActivity.activity_type == activity_type)

        # Order by created_at descending (newest first) and apply pagination
        query = query.order_by(CandidateActivity.created_at.desc()).limit(limit).offset(offset)

        # Execute query
        result = await db.execute(query)
        activities = result.scalars().all()

        # Build response data
        activities_data = []
        for activity in activities:
            activities_data.append({
                "id": str(activity.id),
                "activity_type": activity.activity_type.value,
                "candidate_id": str(activity.candidate_id),
                "vacancy_id": str(activity.vacancy_id) if activity.vacancy_id else None,
                "from_stage": activity.from_stage,
                "to_stage": activity.to_stage,
                "note_id": str(activity.note_id) if activity.note_id else None,
                "tag_id": str(activity.tag_id) if activity.tag_id else None,
                "recruiter_id": str(activity.recruiter_id) if activity.recruiter_id else None,
                "activity_data": activity.activity_data,
                "reason": activity.reason,
                "created_at": activity.created_at.isoformat(),
            })

        # Use the resume_id from filter or first activity if not specified
        response_resume_id = resume_id if resume_id else (
            str(activities[0].candidate_id) if activities else "all"
        )

        response_data = {
            "resume_id": response_resume_id,
            "activities": activities_data,
            "total_count": len(activities_data),
        }

        logger.info(f"Retrieved {len(activities_data)} candidate activities")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except HTTPException:
        raise
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid UUID format provided",
        )
    except Exception as e:
        logger.error(f"Error retrieving candidate activities: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve candidate activities: {str(e)}",
        ) from e


class ActivityTypesResponse(BaseModel):
    """Response model for available activity types."""

    activity_types: List[str] = Field(..., description="List of available activity types")


@router.get(
    "/types",
    response_model=ActivityTypesResponse,
    tags=["Candidate Activities"],
)
async def get_activity_types() -> JSONResponse:
    """
    Get available candidate activity types.

    This endpoint returns a list of all valid activity types that can be used
    for filtering candidate activities.

    Returns:
        JSON response with list of activity types

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/candidate-activities/types")
        >>> response.json()
        {
            "activity_types": [
                "stage_changed",
                "note_added",
                "note_updated",
                "note_deleted",
                "tag_added",
                "tag_removed",
                "ranking_changed",
                "rating_changed",
                "contact_attempt",
                "interview_scheduled",
                "feedback_provided",
                "status_updated"
            ]
        }
    """
    try:
        logger.info("Fetching available activity types")

        activity_types = [t.value for t in CandidateActivityType]

        response_data = {
            "activity_types": activity_types,
        }

        logger.info(f"Retrieved {len(activity_types)} activity types")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error retrieving activity types: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve activity types: {str(e)}",
        ) from e

File: frontend/src/api/candidateActivities.ts
/**
 * Candidate Activities API Client
 *
 * Этот модуль предоставляет клиент для получения истории активности кандидата,
 * включая изменения этапов, добавление/изменение заметок, модификацию тегов
 * и другие значимые события кандидата в течение процесса найма.
 *
 * @example
 * ```ts
 * import { candidateActivitiesClient } from '@/api/candidateActivities';
 *
 * // Получение всех активностей для кандидата
 * const activities = await candidateActivitiesClient.listActivities('resume-123');
 *
 * // Фильтрация активностей по типу
 * const stageChanges = await candidateActivitiesClient.filterByType(
 *   'resume-123',
 *   'stage_changed'
 * );
 *
 * // Получение доступных типов активностей
 * const types = await candidateActivitiesClient.getActivityTypes();
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import type {
  ActivityItem,
  ActivityTimelineResponse,
  ActivityTypesResponse,
  ApiError,
} from '@/types/api';

/**
 * Конфигурация по умолчанию для клиента активностей кандидата
 */
const DEFAULT_CONFIG = {
  baseURL: import.meta.env.VITE_API_URL ?? '',
  timeout: 10000, // 10 секунд
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * Класс клиента API для работы с активностями кандидата
 *
 * Предоставляет методы для получения истории активности кандидата с proper
 * обработкой ошибок и типобезопасностью.
 */
export class CandidateActivitiesClient {
  private client: AxiosInstance;

  /**
   * Создание нового экземпляра клиента активностей кандидата
   *
   * @param config - Опциональные переопределения конфигурации
   */
  constructor(config: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Интерцептор ответов для обработки ошибок
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Преобразование ошибки Axios в стандартизированную ошибку API
   *
   * @param error - Ошибка Axios
   * @returns Преобразованная ошибка API
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Ошибка сети (нет ответа)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Таймаут запроса. Проверьте соединение и попробуйте снова.',
          status: 408,
        };
      }
      return {
        detail: 'Ошибка сети. Проверьте соединение и попробуйте снова.',
        status: 0,
      };
    }

    // Сервер вернул ошибку
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Используем сообщение об ошибке от сервера, если доступно
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Сообщения об ошибках по умолчанию для разных кодов статуса
    const defaultMessages: Record<number, string> = {
      400: 'Неверный запрос. Проверьте введенные данные.',
      401: 'Не авторизован. Войдите в систему.',
      403: 'Доступ запрещен. У вас нет прав для выполнения этого действия.',
      404: 'Ресурс не найден.',
      422: 'Ошибка валидации. Проверьте введенные данные.',
      429: 'Слишком много запросов. Попробуйте позже.',
      500: 'Ошибка сервера. Попробуйте позже.',
      502: 'Ошибка шлюза. Попробуйте позже.',
      503: 'Сервис недоступен. Попробуйте позже.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'Произошла непредвиденная ошибка.',
      status,
    };
  }

  /**
   * Получение списка активностей кандидата с опциональными фильтрами
   *
   * @param resumeId - ID резюме (кандидата)
   * @param activityType - Опциональный фильтр по типу активности
   * @param vacancyId - Опциональный фильтр по ID вакансии
   * @param limit - Максимальное количество активностей для возврата (по умолчанию: 100)
   * @param offset - Количество активностей для пропуска при пагинации (по умолчанию: 0)
   * @returns Список активностей в хронологическом порядке
   * @throws ApiError если получение списка не удалось
   *
   * @example
   * ```ts
   * // Получение всех активностей для кандидата
   * const activities = await candidateActivitiesClient.listActivities('resume-123');
   *
   * // Получение активностей с пагинацией
   * const page1 = await candidateActivitiesClient.listActivities(
   *   'resume-123',
   *   undefined,
   *   undefined,
   *   50,
   *   0
   * );
   *
   * // Фильтрация по вакансии
   * const vacancyActivities = await candidateActivitiesClient.listActivities(
   *   'resume-123',
   *   undefined,
   *   'vacancy-456'
   * );
   * ```
   */
  async listActivities(
    resumeId: string,
    activityType?: string,
    vacancyId?: string,
    limit = 100,
    offset = 0
  ): Promise<ActivityTimelineResponse> {
    try {
      const params: Record<string, string | number> = {
        resume_id: resumeId,
        limit,
        offset,
      };
      if (activityType) params.activity_type = activityType;
      if (vacancyId) params.vacancy_id = vacancyId;

      const response: AxiosResponse<ActivityTimelineResponse> = await this.client.get(
        '/api/candidate-activities/',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Фильтрация активностей кандидата по типу
   *
   * Это удобный метод-обертка для listActivities с фильтром по типу активности.
   *
   * @param resumeId - ID резюме (кандидата)
   * @param activityType - Тип активности для фильтрации (например, 'stage_changed', 'note_added')
   * @param vacancyId - Опциональный фильтр по ID вакансии
   * @param limit - Максимальное количество активностей для возврата (по умолчанию: 100)
   * @param offset - Количество активностей для пропуска при пагинации (по умолчанию: 0)
   * @returns Отфильтрованный список активностей
   * @throws ApiError если фильтрация не удалась
   *
   * @example
   * ```ts
   * // Получение только изменений этапов
   * const stageChanges = await candidateActivitiesClient.filterByType(
   *   'resume-123',
   *   'stage_changed'
   * );
   *
   * // Получение только добавленных заметок
   * const notes = await candidateActivitiesClient.filterByType(
   *   'resume-123',
   *   'note_added'
   * );
   *
   * // Получение добавленных тегов для конкретной вакансии
   * const tags = await candidateActivitiesClient.filterByType(
   *   'resume-123',
   *   'tag_added',
   *   'vacancy-456'
   * );
   * ```
   */
  async filterByType(
    resumeId: string,
    activityType: string,
    vacancyId?: string,
    limit = 100,
    offset = 0
  ): Promise<ActivityTimelineResponse> {
    try {
      const params: Record<string, string | number> = {
        resume_id: resumeId,
        activity_type: activityType,
        limit,
        offset,
      };
      if (vacancyId) params.vacancy_id = vacancyId;

      const response: AxiosResponse<ActivityTimelineResponse> = await this.client.get(
        '/api/candidate-activities/',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Получение доступных типов активностей
   *
   * @returns Список доступных типов активностей, которые можно использовать для фильтрации
   * @throws ApiError если получение не удалось
   *
   * @example
   * ```ts
   * const types = await candidateActivitiesClient.getActivityTypes();
   * console.log(types.activity_types);
   * // [
   * //   'stage_changed',
   * //   'note_added',
   * //   'note_updated',
   * //   'note_deleted',
   * //   'tag_added',
   * //   'tag_removed',
   * //   'ranking_changed',
   * //   'rating_changed',
   * //   'contact_attempt',
   * //   'interview_scheduled',
   * //   'feedback_provided',
   * //   'status_updated'
   * // ]
   * ```
   */
  async getActivityTypes(): Promise<ActivityTypesResponse> {
    try {
      const response: AxiosResponse<ActivityTypesResponse> = await this.client.get(
        '/api/candidate-activities/types'
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Получение базового экземпляра Axios
   *
   * Полезно для выполнения кастомных запросов, не покрытых методами клиента.
   *
   * @returns Экземпляр Axios
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }
}

/**
 * Экземпляр клиента активностей кандидата по умолчанию
 *
 * Используйте этот singleton-экземпляр для всех операций с активностями кандидата.
 */
export const candidateActivitiesClient = new CandidateActivitiesClient();

/**
 * Экспорт класса активностей кандидата для создания кастомных экземпляров
 */
export default CandidateActivitiesClient;
